site/_posts/2024-12-25-rustport.md

```diff
@@ -0,0 +1,320 @@
+using most of our test suite - which is end-to-end tests that run a script or fake a terminal interaction. We would also not have been able to do another C++ release,
+putting some cool improvements into the hands of our users.
+
+Had we chosen to disappear into a hole we might not have finished at all, and we would have to re-do a bunch of work once it became testable.
+We also mostly kept the structure of the C++ code intact - if a function is in the "env" subsystem, it would stay there. Resisting the temptation to
+clean up allowed us to compare the before and after to find places where we had mistranslated something.
+
+So we used [autocxx](https://google.github.io/autocxx/) to generate bindings between C++ and Rust code, allowing us to port one component at a time.
+
+We started[^technically] by porting the builtins. These are essentially little self-contained programs, with their own arguments, streams, exit code, etc.
+That means it's easy to port them separately from the rest of the shell once you have a way to call a Rust builtin from C++, which we had as part of the initial pull request.
+
+Where they connected to the main shell, we used one of three approaches:
+
+1. Add some FFI glue to the C++ to make it callable from Rust, port the caller and leave the callee for later
+2. Move the callee to Rust and, if necessary, make it callable from C++
+3. Write a Rust version of the callee and call it from the ported caller, but leave the C++ version around
+
+For instance, almost every builtin needs to parse its options. We have our own implementation of getopt, that we reimplemented in Rust in the initial PR,
+but the C++ version stuck around until it had no more callers remaining. Otherwise we would have had to write a C++-to-Rust bridge and adjust the C++ callers to use it.
+
+Or the `builtin` builtin needs access to the names of all builtins to print them for `builtin --get-names`. In that case we bridged some access to what amounts to a constant vector of strings in the C++, and eventually moved it over once the users were in Rust.
+
+That's how it went for a while, but we finally hit the more entangled systems, where porting larger chunks felt more productive,
+since that reduced the amount of tricky FFI code to be written only to be thrown away. These were ported in solo efforts.
+This includes the input/output "reader", which is, unsurprisingly, one of fish's biggest parts, ending up at about 13000 lines of Rust.
+
+During the port, we hit a bunch of snags with (auto)cxx. Sometimes it would just not understand a particular C++ construct, and we spent a lot of time trying to figure out ways to please it. As an example, we introduced a struct on the C++ side that wrapped C++'s `vector`, because for some reason autocxx liked to complain about `vector<wstring>`. It lacks support for wstring/wchar, which is understandable because using wchar is a horrible decision - we only do it because it's a historical mistake.
```

Comment:
It's worth mentioning that we ended up forking both `cxx` and `autocxx` to support wchar.

Reply:
Okay, let's phrase it as "We had to fork it to add support" instead of "It lacks support... We forked it to add it".

No real need to dwell on this.